### PR TITLE
mgr/dashboard: host list tables doesn't show all services deployed

### PR DESCRIPTION
**Pull request by avanthakkar**
Cluster->Host Service instances was displaying only the ceph services, but with these changes it'll display instances
of cephadm services as well.

Before:
Screenshot from 2022-07-25 19-44-39

After:
Screenshot from 2022-07-25 18-27-04

Fixes: https://tracker.ceph.com/issues/53210
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

Service instances was displaying only the ceph services, but with these changes it'll display instances
of cephadm services as well.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts [hunk 1]
```diff
@@ -496,12 +497,29 @@ export class HostsComponent extends ListWithDetails implements OnDestroy, OnInit
...
+                });
+              });
+            } else {
+              host['services'].forEach((service: any) => {
+                counts[service.type] = (counts[service.type] || 0) + 1;
+              });
+              host['services'].map((service: any) => {
+                host['service_instances'].add(`${service.type}: ${counts[service.type]}`);
+              });
+            }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:522]: can't we do this process in backend? @avanthakkar

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts [hunk 2]
```diff
@@ -496,12 +497,29 @@ export class HostsComponent extends ListWithDetails implements OnDestroy, OnInit
           hostList.map((host) => {
             const counts = {};
             host['service_instances'] = new Set<string>();
-            host['services'].forEach((service: any) => {
-              counts[service.type] = (counts[service.type] || 0) + 1;
-            });
-            host['services'].map((service: any) => {
-              host['service_instances'].add(`${service.type}: ${counts[service.type]}`);
-            });
+            if (this.orchStatus?.available) {
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:500]: > @avanthakkar Thanks for this PR, But when i tested it in downstream I see services instances are not stable,count value comes and goes off somewhere flaky code causing that issue, Can you please also check the stable view of it?

@sunilangadi2 I can't reproduce can you attach a recording maybe and send me over?

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:522]: @avanthakkar i think we can afford a refactor for this now right? I mean we've been (including myself) have been patching up this area for a while and I think this deserves some cleaning up. Would you mind looking into the cleanup again and give it a try?

**Comment by nizamial09** [comment]
shaman builds [ongoing](https://shaman.ceph.com/builds/ceph/wip-nia-testing-2022-08-01-1145/)

File: src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/01-hosts.e2e-spec.ts [hunk 3]
```diff
@@ -82,5 +82,10 @@ describe('Hosts page', () => {
       const hostname = Cypress._.sample(this.hosts).name;
       hosts.maintenance(hostname, true);
     });
+
+    it('should exit host from maintenance', function () {
+      const hostname = Cypress._.sample(this.hosts).name;
+      hosts.checkServiceInstancesExist(hostname, ['mgr: 1', 'prometheus: 1']);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/01-hosts.e2e-spec.ts:88]: hmm i am not sure if we can expect the prometheus to be in the teuthology branch. normally it is started without monitoring stacks.

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/01-hosts.e2e-spec.ts (hunk 3 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/01-hosts.e2e-spec.ts]: Yeah, I'm testing with cephadm e2e now instead.

File: src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/workflow/08-hosts.e2e-spec.ts [hunk 4]
```diff
@@ -42,4 +42,8 @@ describe('Host Page', () => {
     hosts.add(hostnames[3]);
     hosts.checkExist(hostnames[3], true);
   });
+
+  it('should show the exact count of daemons', function () {
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/workflow/08-hosts.e2e-spec.ts:46]:
```suggestion
  it('should show the exact count of daemons', () {
```

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Review by aaSharma14 — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunilangadi2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:500]: Even i can't reproduce it now.

**Review by sunilangadi2 — APPROVED** [review]
LGTM.

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by nizamial09** [comment]
newly added test is passing and cephadm e2e failure was because of the loki/grafana issue
`✓ should show the exact count of daemons`

**Pull request merged**